src/aws-lambda/mparticle-personalize/mparticle-personalize.js

```diff
@@ -0,0 +1,177 @@
+const personalizeEvents = new AWS.PersonalizeEvents();
+const personalizeRuntime = new AWS.PersonalizeRuntime();
+const axios = require('axios');
+
+exports.handler = async function (event, context) {
+    // Load all of our environment variables from SSM
+    try {
+        let params = {
+            Names: ['/retaildemostore/services_load_balancers/products',
+                    '/retaildemostore/webui/mparticle_s2s_api_key',
+                    '/retaildemostore/webui/mparticle_s2s_secret_key',
+                    'retaildemostore-personalize-event-tracker-id',
+                    'retaildemostore-product-recommendation-campaign-arn'],
+            WithDecryption: false
+        };
+        let responseFromSSM = await SSM.getParameters(params).promise();
+        
+        for(const param of responseFromSSM.Parameters) {
+            if( param.Name === '/retaildemostore/services_load_balancers/products') {
+                var productsServiceURL = param.Value;
+            } else if (param.Name === '/retaildemostore/webui/mparticle_s2s_api_key') {
+                var mpApiKey = param.Value;
+            } else if (param.Name === '/retaildemostore/webui/mparticle_s2s_secret_key') {
+                var mpApiSecret = param.Value;
+            } else if (param.Name === 'retaildemostore-personalize-event-tracker-id') {
+                var personalizeTrackerID = param.Value; 
+            } else if (param.Name === 'retaildemostore-product-recommendation-campaign-arn') {
+                var personalizeCampaignARN = param.Value;
+            }
+        }
+        
+        // Init mParticle libraries for the function invocation
+        var mpApiInstance = new mParticle.EventsApi(new mParticle.Configuration(mpApiKey, mpApiSecret));
+    } catch (e) {
+        console.log("Error getting SSM parameter for loadbalancer.");
+        console.log(e); 
+        throw e;   
+    }
+
+    for (const record of event.Records) {
+        const payloadString = Buffer.from(record.kinesis.data, 'base64').toString('ascii');
+        const payload = JSON.parse(payloadString);
+        const events = payload.events;
+        var amazonPersonalizeUserId;
+        
+        // First, get the mParticle user ID from the events payload.  In this example, mParticle will send all the events
+        // for a particular user in a batch to this lambda.
+        // retreive the mParticle user id which is available for anonymous and known customer profiles
+        var anonymousID = events[0].data.custom_attributes.mpid.toString();
+        
+        // if the customer profile is known then replace the amazon Personalize User id with the actual
+        // personalize Id captured from the user's profile
+        if(payload.user_attributes && payload.user_attributes.amazonPersonalizeId)
+            amazonPersonalizeUserId = payload.user_attributes.amazonPersonalizeId; 
+        else
+            amazonPersonalizeUserId = anonymousID;
```

Comment:
For PutEvents calls for anonymous users, you actually want to exclude userId from the call. Later when the user does identify themselves and events are sent in with the actual userId, Personalize will associate the prior anonymous events with that user. By using the anonymousID here, you will block that from happening and Personalize will consider the events for the anonymous and known user to be for different users.

However, when calling GetRecommendations for an anonymous user, use the sessionId as the userId in the GetRecs call. This will allow you to get personalized recs for anonymous users before the next retraining where the association described above is done.

Reply:
Hi James!

mParticle has the ability to do identity resolution from mParticle's end. Technically, every time a profile is created in mParticle, mParticle will associate the said user with a unique mParticle ID hence the user being tracked is semi-known to our system. Once the user decides to identify themselves within mParticle, we will actually retain the same mParticle ID but just append additional identifiers (e.g.: userId, email, etc..) to the same profile such that any previous information we tracked for the said user will be retained and moving forward, the identified profile will be the profile which will always be associated back to the same user. This is the reason why I have the else clause because that's where I'm linking the MPID we maintain for the said user being tracked my mParticle. Hope this information helps